cranelift/codegen/src/isa/riscv64/inst/imms.rs

```diff
@@ -0,0 +1,237 @@
+        } else {
+            Self { bits: 0 }
+        }
+    }
+
+    /// Create a zero immediate of this format.
+    #[inline(always)]
+    pub fn zero() -> Self {
+        Imm12 { bits: 0 }
+    }
+    #[inline(always)]
+    pub fn as_i16(self) -> i16 {
+        self.bits
+    }
+    #[inline(always)]
+    pub fn as_u32(&self) -> u32 {
+        (self.bits as u32) & 0xfff
+    }
+}
+
+impl Into<i64> for Imm12 {
+    fn into(self) -> i64 {
+        self.bits as i64
+    }
+}
+impl Display for Imm12 {
+    fn fmt(&self, f: &mut Formatter<'_>) -> Result {
+        write!(f, "{:+}", self.bits)
+    }
+}
+
+impl std::ops::Neg for Imm12 {
+    type Output = Self;
+    fn neg(self) -> Self::Output {
+        Self { bits: -self.bits }
+    }
+}
+
+// singed
+#[derive(Clone, Copy, Default)]
+pub struct Imm20 {
+    /// The immediate bits.
+    pub bits: i32,
+}
+
+impl Imm20 {
+    #[inline(always)]
+    pub fn from_bits(bits: i32) -> Self {
+        Self {
+            bits: bits & 0xf_ffff,
+        }
+    }
+    #[inline(always)]
+    pub fn as_u32(&self) -> u32 {
+        (self.bits as u32) & 0xf_ffff
+    }
+}
+
+impl Debug for Imm20 {
+    fn fmt(&self, f: &mut Formatter<'_>) -> Result {
+        write!(f, "{}", self.bits)
+    }
+}
+
+impl Display for Imm20 {
+    fn fmt(&self, f: &mut Formatter<'_>) -> Result {
+        write!(f, "{}", self.bits)
+    }
+}
+
+#[derive(Clone, Copy)]
+pub struct Uimm5 {
+    bits: u8,
+}
+
+impl Uimm5 {
+    pub fn maybe_from_u64(val: u64) -> Option<Uimm5> {
+        if (val >> 5) == 0 {
+            Some(Self {
+                bits: (val & 0b1_1111) as u8,
+            })
+        } else {
+            None
+        }
+    }
+    pub fn from_bits(bits: u8) -> Self {
+        Self { bits }
+    }
+    /// Create a zero immediate of this format.
+    pub fn zero() -> Self {
+        Self { bits: 0 }
+    }
+    pub fn as_u32(&self) -> u32 {
+        (self.bits as u32) & 0b1_1111
+    }
+}
+
+impl Debug for Uimm5 {
+    fn fmt(&self, f: &mut Formatter<'_>) -> Result {
+        write!(f, "{}", self.bits)
+    }
+}
+
+impl Display for Uimm5 {
+    fn fmt(&self, f: &mut Formatter<'_>) -> Result {
+        write!(f, "{}", self.bits)
+    }
+}
+
+/*
+    a imm20 immediate and a Imm12 immediate can generate what immediate.
+    imm20 + imm12
+*/
+pub struct Imm20AndImm12 {}
+impl Imm20AndImm12 {
+    pub(crate) fn min() -> i64 {
+        let imm20_max: i64 = (1 << 19) << 12;
+        let imm12_max = 1 << 11;
+        -imm20_max - imm12_max
+    }
+    pub(crate) fn max() -> i64 {
+        let imm20_max: i64 = ((1 << 19) - 1) << 12;
+        let imm12_max = (1 << 11) - 1;
+        imm20_max + imm12_max
+    }
+    pub(crate) fn generate<R>(
+        value: u64,
+        mut handle_imm: impl FnMut(Option<Imm20>, Option<Imm12>) -> R,
+    ) -> Option<R> {
+        if let Some(imm12) = Imm12::maybe_from_u64(value) {
+            /*
+                can be load using single imm12.
+            */
```

Comment:
Please use line comments instead of block comments.

```suggestion
            // can be load using single imm12.
```